staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go

```diff
@@ -0,0 +1,258 @@
+	// This should be executed after restarting the watch (if applicable) to ensure no file event will be missing.
+	defer ec.queue.Add(workqueueKey)
+
+	// return if file has not removed or renamed.
+	if event.Op&(fsnotify.Remove|fsnotify.Rename) == 0 {
+		return nil
+	}
+
+	if err := watcher.Remove(ec.filePath); err != nil {
+		klog.InfoS("Failed to remove file watch, it may have been deleted", "file", ec.filePath, "err", err)
+	}
+	if err := watcher.Add(ec.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", ec.filePath, err)
+	}
+
+	return nil
+}
+
+// runWorker to process file content
+func (ec *DynamicKMSEncryptionConfigContent) runWorker() {
+	for ec.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem processes file content when there is a message in the queue.
+func (ec *DynamicKMSEncryptionConfigContent) processNextWorkItem() bool {
+	// key here is dummy item in the queue to trigger file content processing.
+	key, quit := ec.queue.Get()
+	if quit {
+		return false
+	}
+	defer ec.queue.Done(key)
+
+	err := ec.loadEncryptionConfig()
+	if err == nil {
+		ec.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// add dummy item back to the queue to trigger file content processing.
+	ec.queue.AddRateLimited(key)
+
+	return true
+}
+
+// loadEncryptionConfig processes the next set of content from the file.
+func (ec *DynamicKMSEncryptionConfigContent) loadEncryptionConfig() error {
+	encryptionConfig, err := ec.readEncryptionFileContent(ec.filePath)
+	if err != nil {
+		return err
+	}
+	if len(encryptionConfig) == 0 {
+		return fmt.Errorf("encryption config file is empty %q", ec.name)
+	}
+
+	// check if encryptionConfig is different from the current. Do nothing if they are the same.
+	if !ec.hasEncryptionConfigChanged(encryptionConfig) {
+		return nil
+	}
+
+	err = ec.processEncryptionConfig(encryptionConfig)
+	if err != nil {
+		return err
+	}
+
+	// update local copy of recent config content only if processing is successful.
+	ec.encryptionConfig.Store(encryptionConfig)
+	klog.V(2).InfoS("Loaded new kms encryption config content", "name", ec.name)
+
+	return nil
+}
+
+func (ec *DynamicKMSEncryptionConfigContent) readEncryptionFileContent(filePath string) ([]byte, error) {
+	return os.ReadFile(ec.filePath)
+}
+
+// processEncryptionConfig processes the encryption config.
+func (ec *DynamicKMSEncryptionConfigContent) processEncryptionConfig(encryptionConfig []byte) error {
+	transformerOverrides, err := encryptionconfig.GetTransformerOverrides(ec.filePath, ec.apiServerDrainCh)
+	if err != nil {
+		klog.Warning("KMS encryption configuration is not valid.", err)
+		return err
+	}
+
+	// update transformers.
+	for groupResource, transformer := range transformerOverrides {
+		ec.storageFactory.UpdateMutableTransformer(groupResource, transformer)
+	}
```

Comment:
This is not enough because it misses the health checks and leaks the old gRPC connections (which you do not notice in tests because they are still cleaned up at shutdown).

Reply:
implemented closing grpc conn.